### PR TITLE
Add  support for feature check.

### PR DESCRIPTION
**Pull request by TuringKi**
This commit implements a very basic feature checking module.

The results like this:

```

../gcc/testsuite/rust/compile/feature.rs:1:12: error: unknown feature 'AA'
    1 | #![feature(AA)] //{ dg-error "unknown feature 'AA'" }
      |            ^

```

This commit is split from pr https://github.com/Rust-GCC/gccrs/pull/1718.

Signed-off-by: Xiao Ma <mxlol233@outlook.com>

### CODE REVIEW

File: gcc/rust/checks/errors/rust-feature.h
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-feature.h:40]: As mentioned on the other PR

1. In my opinion the parameters are not named appropriately and would benefit from a more descriptive name
2. I think we should remove the `gccrs_since` argument and field in the class

File: gcc/rust/checks/errors/rust-feature.cc
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-feature.cc:65]: Is it not possible to avoid the use of macros here? I thought this would be okay with C++'s `constexpr` and so on

**Review by CohenArthur — COMMENTED** [review]
Thanks! This is in a very good state. I have one question regarding the `Feature` class: where is the edition coming from? I'm not aware of Rust features being gated behind an edition, but it might be wrong. Is that directly taken from the rustc implementation?

File: gcc/rust/checks/errors/rust-feature.cc
Comment (philberty) [inline gcc/rust/checks/errors/rust-feature.cc:51]: we need these strings to also have a public enum associated with them then you can use a Feature::ToString for this

**Review by philberty — CHANGES_REQUESTED** [review]
I agree with @CohenArthur we should name our parameters on the Feature constructor here as it helps with readability.

Apart from that the idea here LGTM. There are follow up changes we need to make here.

There is the problem where you use a feature that is not specified in the code and there are features that are allowed by default etc. This scan will only determine the specified features which is the right starting point.

The next PR should be to make the FeatureGate have a function:

```bool featureAvailable (enum Feature);```

So when you do the feature gate check we do it like this:

```
FeatureGate feature_gate;
feature_gate.check (hir);
```

We also need a method:

```bool ok () const;```

this returns true if any errors based on the feature's gates have been emitted.

This along with the public enum will mean we can pass this object as a dependency to the code-generation context so during code-generation we can check if we are compiling something that uses a feature of some kind we can check if its available in the feature gate object. Though we might actually make that a separate error check pass over the crate again but either way it doesn't matter.

**Comment by philberty** [comment]
TLDR i would be happy to merge this PR if you fix:

1. naming of the parameters in the Feature constructor
2. add a new enum for each of the features you have defined so far.

**Comment by philberty** [comment]
Actually after rereading the PR maybe your intention is to also walk the whole crate for features that will require a feature check? I still think the features should be an enum though

**Comment by TuringKi** [comment]
> TLDR i would be happy to merge this PR if you fix:
> 
> 1. naming of the parameters in the Feature constructor
> 2. add a new enum for each of the features you have defined so far.

Thanks for review this pr. I left for Chinese New Year (春节 in chinese). I will check it out about your advice in 1-2 days.

File: gcc/rust/checks/errors/rust-feature.h
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-feature.h:59]: @TuringKi the argument names to the function are still not precise enough IMO
```suggestion
  Feature (Name name, State state, const char *name_str, const char *rustc_since, uint64_t issue_number,
	   const Optional<CompileOptions::Edition> &edition, const char *description)
    : m_state (state), m_name (name), m_name_str (name_str), m_rustc_since (rustc_since), issue (issue_number),
      edition (edition), m_description (description)
  {}
```

**Review by CohenArthur — COMMENTED** [review]
> Thanks for review this pr. I left for Chinese New Year (春节 in chinese). I will check it out about your advice in 1-2 days.

Hope you had a good time :D I've added a suggestion for the arguments name and if you want we can work together on removing the macros and adding @philberty's enum idea

**Comment by TuringKi** [comment]
> > Thanks for review this pr. I left for Chinese New Year (春节 in chinese). I will check it out about your advice in 1-2 days.
> 
> Hope you had a good time :D I've added a suggestion for the arguments name and if you want we can work together on removing the macros and adding @philberty's enum idea

The macro is removed, and the constructor parameter names are changed.

```
enum class Name
  {
    ASSOCIATED_TYPE_BOUNDS,
    INTRINSICS,
    RUSTC_ATTRS,
    DECL_MACRO,
  };
```

I'm not sure this enum definition is ok?

**Review by philberty — APPROVED** [review]
LGTM nice work! I am excited to start walking the crate. And you are right to choose the AST here because the HIR is desugared too much.

**Comment by philberty** [comment]
I am happy with this, its a good starting point! Leaving it over to @CohenArthur to merge when he is happy.

File: gcc/rust/checks/errors/rust-feature-gate.cc
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-feature-gate.cc:52]: Minor GCC coding-style nitpick
```suggestion
		  if (!tname.is_none ())
		      valid_features.push_back (Feature::create (tname.get ()));
		  else
		      rust_error_at (item->get_locus (), "unknown feature '%s'",
				     name.c_str ());
```

File: gcc/rust/checks/errors/rust-feature.cc
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-feature.cc:65]:
```suggestion
  if (Feature::name_hash_map.count (name))
      return Optional<Feature::Name>::some (Feature::name_hash_map.at (name));
      
  return Optional<Feature::Name>::none ();
```
this is a great use of `Optional` btw :) thanks for using it

**Review by CohenArthur — APPROVED** [review]
Great stuff :) I agree that this is a great starting point. Feel free to ignore the style nitpick and I'll fix them in a subsequent PR, or you can fix them now and I'll merge this PR later today. Thanks for working on this!

**Comment by TuringKi** [comment]
The  coding-style  problem is fixed. :)

**Comment by TuringKi** [comment]
> Great stuff :) I agree that this is a great starting point. Feel free to ignore the style nitpick and I'll fix them in a subsequent PR, or you can fix them now and I'll merge this PR later today. Thanks for working on this!

Thanks. I will add "intrinsic" feature gate after this pr is merged.

**Comment by CohenArthur** [comment]
Thanks a lot @TuringKi ! Great job :D

bors r+

**Pull request merged**
